### PR TITLE
MAINT: only add --std=c99 where needed

### PR DESCRIPTION
**Pull request by mattip**
Revert gh-15194, redo gh-14471

~We really only need `--std=c99` for pocketfft compilation, so do not add it unilaterally.~ Add the flag via `setup.py`. As pointed out in scipy/scipy#11270, this also turns on different floating point excess precision handling on 387 FPU.

### CODE REVIEW

**Comment by pv** [comment]
I'd suggest instead to do:
```
def get_build_ext_override():
    from numpy.distutils.commands.build_ext import build_ext as old_build_ext
    class build_ext(old_build_ext):
        def build_extension(self, ext):
            <override compiler flag here>
...

def setup_package():
...
    metadata = dict(
...
        cmdclass={"sdist": sdist_checked,
            'build_ext': get_build_ext_override()},
...
```
to override just the compiler flags used when building the specific extension.

**Comment by mattip** [comment]
The CI run with gcc 4.8 agrees that it is not enough to add the flag for pocketfft and random, we need it for all c-extensions.

**Comment by mattip** [comment]
Turns out we need to override both `build_clib` and `build_ext`

**Comment by mattip** [comment]
the build with gcc 4.8 passes.

**Review by pv — APPROVED** [review]
LGTM.

**Comment by eric-wieser** [comment]
Will this correctly apply the flag for the configtest stuff too?

File: setup.py [hunk 1]
```diff
@@ -225,6 +226,40 @@ def run(self):
...
+    """
+    Custom build commands to add `--std=c99` to compilation
+    """
+    from numpy.distutils.command.build_clib import build_clib
+    from numpy.distutils.command.build_ext import build_ext
+
+    def _is_using_gcc(obj):
+        is_gcc = False
+        if obj.compiler.compiler_type == 'unix':
+            cc = sysconfig.get_config_var("CC")
```
Comment (eric-wieser) [inline setup.py]: Shouldn't we be looking at `obj.compiler` to learn this, rather than global state that may not be equivalent?

Comment (pv) [reply setup.py]: Distutils itself does it like this. But I'm not sure if there is some reason why.

File: setup.py [hunk 2]
```diff
@@ -225,6 +226,40 @@ def run(self):
...
+                args = build_info.get('extra_compiler_args') or []
+                args.append('--std=c99')
+                build_info['extra_compiler_args'] = args
+            build_clib.build_a_library(self, build_info, lib_name, libraries)
+
+    class new_build_ext(build_ext):
+        def build_extension(self, ext):
+            if _is_using_gcc(self):
+                if '--std=c99' not in ext.extra_compile_args:
+                    ext.extra_compile_args.append('--std=c99')
```
Comment (pv) [inline setup.py]: Isn't the option `-std=c99` with single dash?

**Comment by pv** [comment]
It won't apply to configtest I think, as that's done by an earlier config command. Might also be possible to override only the `config` command here.

File: setup.py (hunk 1 above)
Comment (mattip) [reply setup.py]: once burned twice shy: I prefer to mimic distutils as far as possible

File: setup.py (hunk 2 above)
Comment (mattip) [reply setup.py]: fixing

**Comment by charris** [comment]
Thanks Matti. I don't think the config tests need the flag, but it may be worth tracking down the call in case that changes in the future.

**Pull request merged**